texinfo: fix emission of @footnote directives.

marxin opened the pull request:
Right now, labels are emitted as part of a @footnote directive.
That results in e.g.

```
Note1: (@footnote{@w{(1)}
Future versions of GCC may zero-extend...
})
```

which is incorrect and should be rather:

```
Note1: (@footnote{Future versions of GCC may zero-extend...})
```

marxin committed:
texinfo: fix emission of @footnote directives.

Right now, labels are emitted as part of a @footnote directive.
That results in e.g.

Note1: (@footnote{@w{(1)}
Future versions of GCC may zero-extend...
})

which is incorrect and should be rather:

Note1: (@footnote{Future versions of GCC may zero-extend...})

marxin commented:
Texinfo documentation for `@footnote`: https://www.gnu.org/software/texinfo/manual/texinfo/html_node/Footnote-Commands.html

tk0miya commented on sphinx/writers/texinfo.py:
```diff
@@ -1223,7 +1225,10 @@ def depart_sidebar(self, node: Element) -> None:
+        if self.in_footnote:
+            raise nodes.SkipNode
+        else:
+            self.body.append('@w{(')
```
Note: This is still needed for citations.

marxin replied:
All right, so should I add a counter that would increment/decrement once a citation is visited (`in_citation`)?

tk0miya replied:
I'm not sure. Let's think that again if we'll need to support.

tk0miya commented on sphinx/writers/texinfo.py:
```diff
@@ -1223,7 +1225,10 @@ def depart_sidebar(self, node: Element) -> None:
     def visit_label(self, node: Element) -> None:
-        self.body.append('@w{(')
+        if self.in_footnote:
+            raise nodes.SkipNode
```
Could you add a comment why labels are skipped in footnotes?

tk0miya commented on sphinx/writers/texinfo.py:
```diff
@@ -761,10 +761,12 @@ def visit_title_reference(self, node: Element) -> None:
     def visit_paragraph(self, node: Element) -> None:
-        self.body.append('\n')
+        if not self.in_footnote:
+            self.body.append('\n')
```
Is this removal required? It seems Texinfo does not mind if any whitespace exists.

marxin replied:
Well, the generated texinfo output is more readable and easier for writing a test-case. Excluding the hunk, one would see:

```
@chapter The section with a reference to @footnote{
Footnote in section
}
```

tk0miya replied:
Finally, I determined to remove this code to keep our writer simple. I think the simpleness of the writer is more important than the readability of output.

tk0miya commented on tests/test_build_texinfo.py:
```diff
@@ -112,3 +112,12 @@ def test_texinfo_escape_id(app, status, warning):
+    app.builder.build_all()
+
+    output = (app.outdir / 'python.texi').read_text()
+    assert '@footnote{@w{' not in output
```
It's enough to check the form of the new output. Checking the removal of the footnote label is meaningless.

tk0miya reviewed (changes requested):
LGTM with nits.

marxin commented:
@tk0miya May I please remind this pull request where I asked some questions?

tk0miya committed:
Merge branch '4.x' into fix-footnote-in-info

tk0miya committed:
refactor: texinfo: Remove CR char from output

The CR character was added for readability of output.  But it makes the
texinfo writer a bit complicated.  This removes it from output to keep
our code simple (reducing conditions).

tk0miya commented:
Merged. Thank you for your contribution!
